### PR TITLE
#11822 filepath annotations

### LINKED ISSUE #11822: FilePath and friends should have type annotations
Particularly given the text/bytes-mode distinction of how they can work, we should really include type hints with them.

### PR DESCRIPTION
**Pull request by glyph**
## Scope and purpose

Fixes #11822 

## Contributor Checklist:

This process applies to *all* pull requests - no matter how small.
Have a look at [our developer documentation](https://docs.twisted.org/en/latest/core/development/dev-process.html) before submitting your Pull Request.

Below is a non-exhaustive list (as a reminder):

* The title of the PR should describe the changes and starts with the associated issue number, like “#1234 Brief description”.
* A release notes news fragment file was create in src/twisted/newsfragments/ (see: [Release notes fragments docs.](https://docs.twisted.org/en/latest/core/development/dev-process.html#release-notes-management))
* The automated tests were updated.
* Once all checks are green, request a review by leaving a comment that contains exactly the string `please review`.
  Our bot will trigger the review process, by applying the pending review label
  and requesting a review from the Twisted dev team.

### CODE REVIEW

**Comment by glyph** [comment]
review please

**Comment by glyph** [comment]
haven't run the tests yet myself, but the type-check should be clean

File: src/twisted/python/filepath.py [hunk 1]
```diff
@@ -38,7 +39,24 @@
     S_IXOTH,
     S_IXUSR,
 )
-from typing import IO, Union, cast
+from typing import (
+    IO,
+    TYPE_CHECKING,
+    AnyStr,
+    Callable,
+    Dict,
```
Comment (exarkun) [inline src/twisted/python/filepath.py:48]: Why `__future__.annotations` _and_ `Dict`, `List`, etc?

Comment (glyph) [reply src/twisted/python/filepath.py:48]: `__future__.annotations` lets you forward-declare stuff, it doesn't affect builtin types.  Which means that APIs like [`inspect.get_annotations(..., eval_str=True)`](https://docs.python.org/3.11/library/inspect.html#inspect.get_annotations) (and its more ad-hoc predecessors) will work just fine in the presence of `__future__.annotations` on older Python versions, but will *not* work if you start using `list[...]`, `dict[...]`, et. al.

Comment (glyph) [reply src/twisted/python/filepath.py:48]: (When we drop <3.9, [we can ask pyupgrade to make this transformation for us mechanically](https://github.com/asottile/pyupgrade/tree/67e006e96f7fab6533ab5f831872e72c109a7b28#pep-585-typing-rewrites), and I look forward to doing so.)

**Comment by glyph** [comment]
So, those test failures were embarrassing. Let's try that again…

**Comment by glyph** [comment]
Looks like I've got a few minor coverage gaps which should be addressed but otherwise I think we're good now

**Comment by glyph** [comment]
OK. Coverage gaps all addressed.

**Review by adiroiban — APPROVED** [review]
I don't have much experience with Python typing.
I did a very quick look and it looks ok
I hope @graingert can also take another look to check all is ok.

Since this is waiting in the queue since March, as long as the tests are green, I think that this can be  merged.

Thanks Glyph for the update.

**Comment by glyph** [comment]
thanks @adiroiban ! I'm happy to take any post-hoc feedback from @graingert (or anyone else for that matter) but as you note, this *has* been waiting since March, so let's get it out of the queue and move things along, there are plenty of other PRs that need review attention…

File: src/twisted/python/zippath.py [hunk 2]
```diff
@@ -206,11 +270,13 @@ class ZipArchive(ZipPath):
     converting if required.
     """
 
+    _zipfileFilename: AnyStr
+
     @property
     def archive(self):
         return self
```
Comment (graingert) [inline src/twisted/python/zippath.py:277]: this is the only method not annotated, it should probably be  `-> Self: # type: ignore[override]`

File: src/twisted/python/filepath.py [hunk 3]
```diff
@@ -273,17 +283,87 @@ def _secureEnoughString(path):
     return _coerceToFilesystemEncoding(path, secureishString)
```
Comment (graingert) [inline src/twisted/python/filepath.py:283]: this is missing a type annotation - should be `def _secureEnoughString(path: AnyStr) -> AnyStr: ...`

**Pull request merged**